docs/articles/documentation/test-api/obtaining-data-from-the-client.md

```diff
@@ -0,0 +1,200 @@
+
+The following example shows how to create a client function.
+
+```js
+import { ClientFunction } from 'testcafe';
+
+const getWindowLocation = ClientFunction(() => window.location);
+```
+
+## Executing Client Functions
+
+To execute a client function, call it with the `await` keyword.
+
+```js
+import { ClientFunction } from 'testcafe';
+
+const getWindowLocation = ClientFunction(() => window.location);
+
+fixture `My fixture`
+    .page('http://www.example.com/');
+
+test('My Test', async t => {
+    const location = await getWindowLocation();
+});
+```
+
+## Options
+
+You can pass the following options to the
+[ClientFunction constructor](#creating-client-functions) or the
+[t.eval](#one-time-client-code-execution) function.
+  
+### options.dependencies
+
+**Type**: Object
+
+Contains functions, variables or objects used by the client function internally.
+Properties of the `dependencies` object will be added to the client function's scope as variables.
+
+The following sample demonstrates a client function (`getElemWidth`) that
+internally calls a [selector](selecting-page-elements/selectors.md) (`getElemById`).
+This selector is passed to `getElemWidth` as a dependency.
+
+```js
+import { Selector, ClientFunction } from 'testcafe';
+
+const getElemById  = Selector(id => document.getElementById(id));
+const getElemWidth = ClientFunction(id => getElemById(id).width, {
+     dependencies: { getElemById }
+});
+```
+
+> When a client function calls a [selector](selecting-page-elements/selectors.md) internally,
+> the selector does not wait for the element to appear in the DOM
+> but is executed at once, like a client function.
+
+### options.boundTestRun
+
+**Type**: Object
+
+If you need to call a client function from a Node.js callback,
+assign the current [test controller](test-code-structure.md#test-controller)
+to the `boundTestRun` option.
+
+For details, see [Calling Client Functions from Node.js Callbacks](#calling-client-functions-from-nodejs-callbacks).
+
+### Overwriting Options
+
+You can overwrite client function options by using the ClientFunction's `with` function.
+
+```text
+clientFunction.with(options) → ClientFunction
+```
+
+`with` returns a new client function with a different set of options that includes options
+from the original function and new `options` that overwrite the original ones.
+
+The sample below shows how to overwrite the client function options to use a different selector.
+
+```js
+import { Selector, ClientFunction } from 'testcafe';
+
+const getThirdElemByClass = Selector(cl => document.getElementsByClassName(cl), { index: 2 });
+const getThirdElemWidth   = ClientFunction(cl => getThirdElemByClass(cl).width, {
+     dependencies: { getThirdElemByClass }
+});
+
+const getFourthElemByClass  = getThirdElemByClass.with({ index: 3 });
+const getFourthElemWidth    = getThirdElemWidth.with({
+    dependencies: { getFourthElemByClass }
+});
```

Comment:
It will not work because `getThirdElemWidth` Client Function required `getThirdElemByClass` by dependencies. We need refactor this example.